bindings/kitsune_client/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+                        log::error!("failure to query peer store: {err}");
+                        continue;
+                    }
+                };
+                if maybe_agent.is_some() {
+                    break;
+                }
+            }
+        })
+        .await
+        .context("failure to join space")
+    }
+
+    #[wind_tunnel_instrument]
+    pub async fn say(&self, message: &str) -> anyhow::Result<Vec<OpId>> {
+        let state = self.state.lock().unwrap();
+        let message_op = MemoryOp::new(Timestamp::now(), message.into());
+        let message_ids = state
+            .space
+            .op_store()
+            .process_incoming_ops(vec![message_op.into()])
+            .await
+            .context("failure to write ops to the store")?;
+        state
+            .space
+            .inform_ops_stored(
+                message_ids
+                    .clone()
+                    .into_iter()
+                    .map(|message_id| StoredOp {
+                        created_at: Timestamp::now(),
+                        op_id: message_id,
+                    })
+                    .collect(),
+            )
+            .await?;
+        log::info!("agent {} said {}", state.agent.agent(), message);
+
+        self.reporter.add_custom(
+            ReportMetric::new("message said")
+                .with_tag("agent_id", state.agent.agent().to_string())
+                .with_field("message_said", message_ids.len() as u32),
+        );
+
+        Ok(message_ids)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use kitsune2_bootstrap_srv::{BootstrapSrv, Config};
+    use sbd_server::SbdServer;
+    use std::time::Duration;
+    use wind_tunnel_core::prelude::ShutdownHandle;
+    use wind_tunnel_instruments::{ReportConfig, Reporter};
+
+    pub(crate) fn test_reporter() -> Arc<Reporter> {
+        let runtime = tokio::runtime::Handle::current();
+        let shutdown_listener = ShutdownHandle::new().new_listener();
+        Arc::new(
+            ReportConfig::new("".to_string(), "".to_string())
+                .enable_in_memory()
+                .init_reporter(&runtime, shutdown_listener)
+                .unwrap(),
+        )
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn say_something_to_other_chatter() {
+        env_logger::init();
+        let signal_server = SbdServer::new(Arc::new(sbd_server::Config {
+            bind: vec!["127.0.0.1:0".to_string()],
+            ..Default::default()
+        }))
+        .await
+        .unwrap();
+        // let signal_server = tokio::task::spawn(
+        //     SbdServer::new(Arc::new(sbd_server::Config {
+        //         bind: vec!["127.0.0.1:0".to_string()],
+        //         ..Default::default()
+        //     })), // .await
+        // )
+        // .await
+        // .unwrap()
+        // .unwrap();
+        let signal_server_url = format!("ws://localhost:{}", signal_server.bind_addrs()[0].port());
```

Comment:
could `localhost` potentially resolve the IPv6 variant `::1`?

```suggestion
        let signal_server_url = format!("ws://127.0.0.1:{}", signal_server.bind_addrs()[0].port());
```